chore(css): Move CSS examples - Feature queries, overflow, subgrid, masonry (#36693)
mdn/content: files/en-us/web/css/css_flow_layout/flow_layout_and_overflow/index.md

{{CSSRef}}

When there is more content than can fit into a container, an overflow situation occurs. Understanding how overflow behaves is important in dealing with any element with a constrained size in CSS. This guide explains how overflow works when working with normal flow.
The HTML is the same in each example, so it's visible in the first section, and hidden in others for brevity.

## What is overflow?

Giving an element a fixed height and width, then adding significant content to the box, creates a basic overflow example:

```html live-sample___overflow
<div class="box">
  <p>
    One November night in the year 1782, so the story runs, two brothers sat
    over their winter fire in the little French town of Annonay, watching the
    grey smoke-wreaths from the hearth curl up the wide chimney.
  </p>
</div>
<p>

>   Their names were Stephen and Joseph Montgolfier, they were papermakers by

Reviewer: ```suggestion
  Their names were Stephen and Joseph Montgolfier. They were papermakers by
```

  any special inspiration from the fact.
</p>
```

```css live-sample___overflow
body {
  font: 1.2em / 1.5 sans-serif;
}

.box {
  width: 300px;
  height: 100px;
  border: 5px solid rebeccapurple;
  padding: 10px;
}
```

{{EmbedLiveSample("overflow", "", "370px")}}

The content goes into the box. Once it fills the box, it continues to overflow in a visible way, displaying content outside the box, potentially displaying under subsequent content. The property that controls how overflow behaves is the [`overflow`](/en-US/docs/Web/CSS/overflow) property which has an initial value of `visible`. This is why we can see the overflow content.

## Controlling overflow

There are other values that control how overflow content behaves. To hide overflowing content use a value of `hidden`. This may cause some of your content to not be visible.

```html hidden live-sample___hidden
<div class="box">
  <p>
    One November night in the year 1782, so the story runs, two brothers sat
    over their winter fire in the little French town of Annonay, watching the
    grey smoke-wreaths from the hearth curl up the wide chimney.
  </p>
</div>
<p>

>   Their names were Stephen and Joseph Montgolfier, they were papermakers by

Reviewer: ```suggestion
  Their names were Stephen and Joseph Montgolfier. They were papermakers by
```

  trade, and were noted as possessing thoughtful minds and a deep interest in
  all scientific knowledge and new discovery.
</p>
<p>
  Before that night—a memorable night, as it was to prove—hundreds of millions
  of people had watched the rising smoke-wreaths of their fires without drawing
  any special inspiration from the fact.
</p>
```

```css live-sample___hidden
body {
  font: 1.2em / 1.5 sans-serif;
}
.box {
  width: 300px;
  height: 100px;
  border: 5px solid rebeccapurple;
  padding: 10px;
  overflow: hidden;
}
```

{{EmbedLiveSample("hidden", "", "370px")}}

Using a value of `scroll` contains the content in its box and add scrollbars to enable viewing it. Scrollbars will be added even if the content fits in the box.

```html hidden live-sample___scroll
<div class="box">
  <p>
    One November night in the year 1782, so the story runs, two brothers sat
    over their winter fire in the little French town of Annonay, watching the
    grey smoke-wreaths from the hearth curl up the wide chimney.
  </p>
</div>
<p>

>   Their names were Stephen and Joseph Montgolfier, they were papermakers by
>   trade, and were noted as possessing thoughtful minds and a deep interest in
>   all scientific knowledge and new discovery.

Reviewer: ```suggestion
Their names were Stephen and Joseph Montgolfier; they were papermakers by trade and were noted as possessing thoughtful minds and a deep interest in all scientific knowledge and new discoveries.
```
per Grammarly ;)

```html hidden live-sample___overflow-y
<div class="box">
  <p>
    One November night in the year 1782, so the story runs, two brothers sat
    over their winter fire in the little French town of Annonay, watching the
    grey smoke-wreaths from the hearth curl up the wide chimney.
  </p>
</div>
<p>
  Their names were Stephen and Joseph Montgolfier, they were papermakers by
  trade, and were noted as possessing thoughtful minds and a deep interest in
  all scientific knowledge and new discovery.
</p>
<p>
  Before that night—a memorable night, as it was to prove—hundreds of millions
  of people had watched the rising smoke-wreaths of their fires without drawing
  any special inspiration from the fact.
</p>
```

```css live-sample___overflow-y
body {
  font: 1.2em / 1.5 sans-serif;
}
.box {
  width: 300px;
  height: 100px;
  border: 5px solid rebeccapurple;
  padding: 10px;
  overflow-y: scroll;
}
```

{{EmbedLiveSample("overflow-y", "", "370px")}}

## Flow Relative Properties

> In the guide to [Writing Modes and Flow Layout](/en-US/docs/Web/CSS/CSS_flow_layout/Flow_layout_and_writing_modes), we looked at the `block-size` and `inline-size` properties, which make more sense when working with different writing modes than tying our layout to the physical dimensions of the screen. The Level 3 Overflow Module also includes flow relative properties for overflow - [`overflow-block`](/en-US/docs/Web/CSS/@media/overflow-block) and [`overflow-inline`](/en-US/docs/Web/CSS/@media/overflow-inline). These correspond to `overflow-x` and `overflow-y` but the mapping depends on the writing mode of the document.

Reviewer: ```suggestion
In the guide to [Writing Modes and Flow Layout](/en-US/docs/Web/CSS/CSS_flow_layout/Flow_layout_and_writing_modes), we looked at the `block-size` and `inline-size` properties, which make more sense when working with different writing modes than tying our layout to the physical dimensions of the screen. The [CSS overflow module](/en-US/docs/Web/CSS/CSS_overflow) also includes flow relative properties for overflow - [`overflow-block`](/en-US/docs/Web/CSS/@media/overflow-block) and [`overflow-inline`](/en-US/docs/Web/CSS/@media/overflow-inline). These correspond to `overflow-x` and `overflow-y` but the mapping depends on the writing mode of the document.
```

## Indicating Overflow

> In the Level 3 Overflow specification, there are some properties which can help improve the way content looks in an overflow situation.

Reviewer: ```suggestion
In the CSS overflow module, there are some properties that can help improve the way content looks in an overflow situation.
```

This is useful in the situation where you have a listing of articles, for example, and you display the listings in fixed height boxes which only take a limited amount of text. It may not be obvious to the reader that there is more content to click through to when clicking the box or the title. An ellipsis indicates clearly the fact there is more content. The specification would allow a string of content or a regular ellipsis to be inserted.

## Summary

Whether you are in continuous media on the web or in a Paged Media format such as print or EPUB, understanding how content overflows is useful when dealing with any layout method. By understanding how overflow works in normal flow, you should find it easier to understand the implications of overflow content in layout methods such as grid and flexbox.

## See also

- [Overflowing content](/en-US/docs/Learn/CSS/Building_blocks/Overflowing_content) guide

> - [CSS overflow](/en-US/docs/Web/CSS/CSS_overflow)
> - [CSS containment](/en-US/docs/Web/CSS/CSS_containment)

Reviewer: ```suggestion
- [CSS overflow](/en-US/docs/Web/CSS/CSS_overflow) module
- [CSS containment](/en-US/docs/Web/CSS/CSS_containment) module
```
